### ISSUE TITLE
issue in provisioning deviceplugin on debian stretch when docker-ce is pinned to 17.03.2

### ISSUE THREAD

**Issue by suneeta-mall** [kubernetes/kops#6324]
**1. What `kops` version are you running? The command `kops version`, will display
 this information.**
kops Version 1.11.0

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**
Kubernetes version: v1.12.4

**3. What cloud provider are you using?**
AWS

**4. What commands did you run?  What is the simplest way to reproduce this issue?**
@dcwangmit01 did [this device plugin hook for AWS](https://github.com/kubernetes/kops/tree/master/hooks/nvidia-device-plugin). In [nvidia docker2 install step](https://github.com/kubernetes/kops/blob/master/hooks/nvidia-device-plugin/image/files/02-nvidia-docker.sh#L69), the script install 
```
apt-get install -y --allow-downgrades -o Dpkg::Options::="--force-confold" \
  nvidia-docker2 \
  nvidia-container-runtime \
  docker-ce
```
This has worked until 9th Jan 2019! I dont have the full details, but some recent changes in nvidia docker seem to fail this install of latest kops stable image `kope.io/k8s-1.11-debian-stretch-amd64-hvm-ebs-2018-08-17`  with error 
```
The following packages have unmet dependencies:
 nvidia-docker2 : Depends: docker-ce (= 5:18.09.0~3-0~debian-stretch) but 5:18.09.1~3-0~debian-stretch is to be installed
E: Unable to correct problems, you have held broken packages.
```

because kops nodeup pins the docker version to `17.03.2`. 
I managed to pass the dependency issue by using these specific version of nvidia-docker2 and nvidia-runtime 

```
apt-get install -y --allow-downgrades -o Dpkg::Options::=--force-confold nvidia-docker2=2.0.3+docker17.03.2-1 nvidia-container-runtime=2.0.0+docker17.03.2-1 docker-ce=17.03.2~ce-0~debian-stretch
```

from available, see below:

```
$ apt-cache policy nvidia-docker2
nvidia-docker2:
  Installed: 2.0.3+docker17.03.2-1
  Candidate: 2.0.3+docker18.09.0-1
  Version table:
     2.0.3+docker18.09.0-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
     2.0.3+docker18.06.1-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
     2.0.3+docker18.06.0-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
     2.0.3+docker18.03.1-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
     2.0.3+docker18.03.0-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
     2.0.3+docker17.12.1-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
     2.0.3+docker17.12.0-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
     2.0.3+docker17.09.1-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
     2.0.3+docker17.09.0-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
     2.0.3+docker17.06.2-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
 *** 2.0.3+docker17.03.2-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
        100 /var/lib/dpkg/status
     2.0.2+docker17.12.0-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
     2.0.2+docker17.09.1-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
     2.0.2+docker17.09.0-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
     2.0.2+docker17.06.2-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
     2.0.2+docker17.03.2-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
     2.0.1+docker17.09.1-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
     2.0.1+docker17.09.0-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
     2.0.1+docker17.06.2-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
     2.0.1+docker17.03.2-1 500
        500 https://nvidia.github.io/nvidia-docker/debian9/amd64  Packages
````
Any other later version than `2.0.3+docker17.03.2-1`  seem to look for docker-ce-18.

The install problem seem to have bypassed but it seems that GPU pods are never provisioned anymore. 
Node describe does not show any GPU detected as well:
```
kubectl describe node ip-PIP.ap-southeast-2.compute.internal
Name:               ip-PIP.ap-southeast-2.compute.internal
Roles:              node
Labels:             beta.kubernetes.io/arch=amd64
                    beta.kubernetes.io/instance-type=p2.xlarge
                    beta.kubernetes.io/os=linux
                    failure-domain.beta.kubernetes.io/region=ap-southeast-2
                    failure-domain.beta.kubernetes.io/zone=ap-southeast-2c
                    gpu=true
                    kops.k8s.io/instancegroup=spotgpu
                    kubernetes.io/hostname=ip-PIP.ap-southeast-2.compute.internal
                    kubernetes.io/role=node
                    node-role.kubernetes.io/node=
                    spot=true
Annotations:        flannel.alpha.coreos.com/backend-data: {"VtepMAC":"26:29:c4:06:94:51"}
                    flannel.alpha.coreos.com/backend-type: vxlan
                    flannel.alpha.coreos.com/kube-subnet-manager: true
                    flannel.alpha.coreos.com/public-ip: 172.20.106.138
                    node.alpha.kubernetes.io/ttl: 0
                    volumes.kubernetes.io/controller-managed-attach-detach: true
CreationTimestamp:  Fri, 11 Jan 2019 13:40:38 +1100
Taints:             <none>
Unschedulable:      false
Conditions:
  Type             Status  LastHeartbeatTime                 LastTransitionTime                Reason                       Message
  ----             ------  -----------------                 ------------------                ------                       -------
  OutOfDisk        False   Fri, 11 Jan 2019 14:10:25 +1100   Fri, 11 Jan 2019 13:40:38 +1100   KubeletHasSufficientDisk     kubelet has sufficient disk space available
  MemoryPressure   False   Fri, 11 Jan 2019 14:10:25 +1100   Fri, 11 Jan 2019 13:40:38 +1100   KubeletHasSufficientMemory   kubelet has sufficient memory available
  DiskPressure     False   Fri, 11 Jan 2019 14:10:25 +1100   Fri, 11 Jan 2019 13:40:38 +1100   KubeletHasNoDiskPressure     kubelet has no disk pressure
  PIDPressure      False   Fri, 11 Jan 2019 14:10:25 +1100   Fri, 11 Jan 2019 13:40:38 +1100   KubeletHasSufficientPID      kubelet has sufficient PID available
  Ready            True    Fri, 11 Jan 2019 14:10:25 +1100   Fri, 11 Jan 2019 13:45:23 +1100   KubeletReady                 kubelet is posting ready status
Addresses:
  InternalIP:   -
  ExternalIP:   -
  InternalDNS:  ip-PIP.ap-southeast-2.compute.internal
  Hostname:     ip-PIP.ap-southeast-2.compute.internal
  ExternalDNS:  ec2-EIP.ap-southeast-2.compute.amazonaws.com
Capacity:
 attachable-volumes-aws-ebs:  39
 cpu:                         4
 ephemeral-storage:           125753328Ki
 hugepages-1Gi:               0
 hugepages-2Mi:               0
 memory:                      62884028Ki
 pods:                        110
Allocatable:
 attachable-volumes-aws-ebs:  39
 cpu:                         4
 ephemeral-storage:           115894266893
 hugepages-1Gi:               0
 hugepages-2Mi:               0
 memory:                      62781628Ki
 pods:                        110
System Info:
 Kernel Version:             4.9.0-7-amd64
 OS Image:                   Debian GNU/Linux 9 (stretch)
 Operating System:           linux
 Architecture:               amd64
 Container Runtime Version:  docker://17.3.2
 Kubelet Version:            v1.12.4
 Kube-Proxy Version:         v1.12.4
PodCIDR:                     100.96.34.0/24
ProviderID:                  aws:///ap-southeast-2c/i-0ee71bf2fdfe8a1de
Non-terminated Pods:         (4 in total)
  Namespace                  Name                                                            CPU Requests  CPU Limits  Memory Requests  Memory Limits  AGE
  ---------                  ----                                                            ------------  ----------  ---------------  -------------  ---
  default                    datadog-agent-bd7dfd96d-sdwrp                                   200m (5%)     200m (5%)   256Mi (0%)       256Mi (0%)     25m
  kube-system                kube-flannel-ds-jkdcg                                           100m (2%)     0 (0%)      100Mi (0%)       100Mi (0%)     29m
  kube-system                kube-proxy-ip-PIP.ap-southeast-2.compute.internal    100m (2%)     0 (0%)      0 (0%)           0 (0%)         29m
  kube-system                nvidia-device-plugin-daemonset-1.12-6qvgj                       0 (0%)        0 (0%)      0 (0%)           0 (0%)         29m
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource                    Requests    Limits
  --------                    --------    ------
  cpu                         400m (10%)  200m (5%)
  memory                      356Mi (0%)  356Mi (0%)
  ephemeral-storage           0 (0%)      0 (0%)
  attachable-volumes-aws-ebs  0           0
Events:
  Type    Reason                   Age                From                                                           Message
  ----    ------                   ----               ----                                                           -------
  Normal  Starting                 29m                kubelet, ip-PIP.ap-southeast-2.compute.internal     Starting kubelet.
  Normal  NodeHasSufficientDisk    29m (x2 over 29m)  kubelet, ip-PIP.ap-southeast-2.compute.internal     Node ip-PIP.ap-southeast-2.compute.internal status is now: NodeHasSufficientDisk
  Normal  NodeHasSufficientMemory  29m (x2 over 29m)  kubelet, ip-PIP.ap-southeast-2.compute.internal     Node ip-PIP.ap-southeast-2.compute.internal status is now: NodeHasSufficientMemory
  Normal  NodeHasNoDiskPressure    29m (x2 over 29m)  kubelet, ip-PIP.ap-southeast-2.compute.internal     Node ip-PIP.ap-southeast-2.compute.internal status is now: NodeHasNoDiskPressure
  Normal  NodeHasSufficientPID     29m (x2 over 29m)  kubelet, ip-PIP.ap-southeast-2.compute.internal     Node ip-PIP.ap-southeast-2.compute.internal status is now: NodeHasSufficientPID
  Normal  NodeAllocatableEnforced  29m                kubelet, ip-PIP.ap-southeast-2.compute.internal     Updated Node Allocatable limit across pods
  Normal  Starting                 29m                kube-proxy, ip-PIP.ap-southeast-2.compute.internal  Starting kube-proxy.
  Normal  NodeReady                29m                kubelet, ip-PIP.ap-southeast-2.compute.internal     Node ip-PIP.ap-southeast-2.compute.internal status is now: NodeReady
  Normal  Starting                 25m                kubelet, ip-PIP.ap-southeast-2.compute.internal     Starting kubelet.
  Normal  NodeHasSufficientDisk    25m                kubelet, ip-PIP.ap-southeast-2.compute.internal     Node ip-PIP.ap-southeast-2.compute.internal status is now: NodeHasSufficientDisk
  Normal  NodeHasSufficientMemory  25m                kubelet, ip-PIP.ap-southeast-2.compute.internal     Node ip-PIP.ap-southeast-2.compute.internal status is now: NodeHasSufficientMemory
  Normal  NodeHasNoDiskPressure    25m                kubelet, ip-PIP.ap-southeast-2.compute.internal     Node ip-PIP.ap-southeast-2.compute.internal status is now: NodeHasNoDiskPressure
  Normal  NodeHasSufficientPID     25m                kubelet, ip-PIP.ap-southeast-2.compute.internal     Node ip-PIP.ap-southeast-2.compute.internal status is now: NodeHasSufficientPID
  Normal  NodeNotReady             25m                kubelet, ip-PIP.ap-southeast-2.compute.internal     Node ip-PIP.ap-southeast-2.compute.internal status is now: NodeNotReady
  Normal  NodeAllocatableEnforced  25m                kubelet, ip-PIP.ap-southeast-2.compute.internal     Updated Node Allocatable limit across pods
  Normal  NodeReady                25m                kubelet, ip-PIP.ap-southeast-2.compute.internal     Node ip-PIP.ap-southeast-2.compute.internal status is now: NodeReady
```
Device plugin installed here found here https://raw.githubusercontent.com/NVIDIA/k8s-device-plugin/v1.12/nvidia-device-plugin.yml 

**5. What happened after the commands executed?**
Allocatable GPUs are not exposed. 

**6. What did you expect to happen?**
GPUs allocability should be exposed and pods should be provisioned. The same setup have been working well but a recent I presume nvidia-docker release has broken things. 

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**
```
apiVersion: kops/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: 2018-12-27T03:38:29Z
  name: mycluster.com
spec:
  api:
    dns: {}
  authorization:
    alwaysAllow: {}
  channel: stable
  cloudLabels:
    k8s.io/cluster-autoscaler/enabled: ""
    kubernetes.io/cluster/mycluster.com: ""
  cloudProvider: aws
  configBase: s3://mycluster.com-state/mycluster.com
  dnsZone: MYDNZ
  etcdClusters:
  - etcdMembers:
    - encryptedVolume: true
      instanceGroup: master-ap-southeast-2a
      name: a
    - encryptedVolume: true
      instanceGroup: master-ap-southeast-2b
      name: b
    - encryptedVolume: true
      instanceGroup: master-ap-southeast-2c
      name: c
    name: main
  - etcdMembers:
    - encryptedVolume: true
      instanceGroup: master-ap-southeast-2a
      name: a
    - encryptedVolume: true
      instanceGroup: master-ap-southeast-2b
      name: b
    - encryptedVolume: true
      instanceGroup: master-ap-southeast-2c
      name: c
    name: events
  fileAssets:
  - content: |
      [Unit]
      Description=Limited resources slice for Kubernetes services
      Documentation=man:systemd.special(7)
      DefaultDependencies=no
      Before=slices.target
      Requires=-.slice
      After=-.slice
    name: podruntime-slice
    path: /usr/lib/systemd/system/podruntime.slice
  iam:
    allowContainerRegistry: true
    legacy: false
  kubeControllerManager:
    horizontalPodAutoscalerUseRestClients: true
  kubelet:
    anonymousAuth: false
  kubernetesApiAccess:
  - MYCIDR
  kubernetesVersion: 1.12.4
  masterInternalName: api.internal.mycluster.com
  masterKubelet:
    kubeletCgroups: /podruntime.slice
    runtimeCgroups: /podruntime.slice
  masterPublicName: api.mycluster.com
  networkCIDR: 172.20.0.0/16
  networking:
    flannel:
      backend: vxlan
  nonMasqueradeCIDR: 100.64.0.0/10
  sshAccess:
  - MYCIDR
  subnets:
  - cidr: 172.20.32.0/19
    name: ap-southeast-2a
    type: Public
    zone: ap-southeast-2a
  - cidr: 172.20.64.0/19
    name: ap-southeast-2b
    type: Public
    zone: ap-southeast-2b
  - cidr: 172.20.96.0/19
    name: ap-southeast-2c
    type: Public
    zone: ap-southeast-2c
  topology:
    dns:
      type: Public
    masters: public
    nodes: public

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-12-27T03:38:29Z
  labels:
    kops.k8s.io/cluster: mycluster.com
  name: master-ap-southeast-2a
spec:
  image: kope.io/k8s-1.11-debian-stretch-amd64-hvm-ebs-2018-08-17
  machineType: m4.xlarge
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-ap-southeast-2a
  role: Master
  subnets:
  - ap-southeast-2a

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-12-27T03:38:29Z
  labels:
    kops.k8s.io/cluster: mycluster.com
  name: master-ap-southeast-2b
spec:
  image: kope.io/k8s-1.11-debian-stretch-amd64-hvm-ebs-2018-08-17
  machineType: m4.xlarge
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-ap-southeast-2b
  role: Master
  subnets:
  - ap-southeast-2b

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-12-27T03:38:29Z
  labels:
    kops.k8s.io/cluster: mycluster.com
  name: master-ap-southeast-2c
spec:
  image: kope.io/k8s-1.11-debian-stretch-amd64-hvm-ebs-2018-08-17
  machineType: m4.xlarge
  maxSize: 1
  minSize: 1
  nodeLabels:
    kops.k8s.io/instancegroup: master-ap-southeast-2c
  role: Master
  subnets:
  - ap-southeast-2c

---

apiVersion: kops/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: 2018-12-27T03:54:34Z
  labels:
    kops.k8s.io/cluster: mycluster.com
  name: nodes
spec:
  hooks:
  - execContainer:
      image: suneetamall/nvidia-device-plugin:0.1.0
  image: kope.io/k8s-1.11-debian-stretch-amd64-hvm-ebs-2018-08-17
  machineType: p2.xlarge
  maxPrice: "0.504"
  maxSize: 1
  minSize: 0
  nodeLabels:
    kops.k8s.io/instancegroup: nodes
  role: Node
  rootVolumeOptimization: true
  subnets:
  - ap-southeast-2a
  - ap-southeast-2c

```
**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**

**9. Anything else do we need to know?**
I do this logs provisioning `I0111 02:45:23.527850   10387 nvidia.go:100] NVML initialized. Number of nvidia devices: 1` which seem to imply that drivers are setup correctly and Kubernetes components seem to work. Just the deviceplugin bit seems broken. I dont have much information on how to fix this at the moment so I will appreciate any help. 
Having said that, in kops hook, we should pin the nvidia docker version as well (just as we pin docker-ce in kops)

**Comment by suneeta-mall** [comment]
The nvidia-docker compatibility issue tracked here https://github.com/NVIDIA/nvidia-docker/issues/903  seems relevant

**Comment by dcwangmit01** [comment]
Nice debugging.  This guy says nvidia-docker packages will be regenerated Monday.

https://github.com/NVIDIA/nvidia-docker/issues/902#issuecomment-453375926

**Comment by suneeta-mall** [comment]
@dcwangmit01 Yeah, they said they will regenerate.

I raised the bug because IMO we should pin the nvidia-docker2 and runtime to specific release - pulling in latest is likely to cause issues like these in production. 

I managed to sort the issue out have mix of packer scripts and exeContainer hook based on your that works for now. Given that I will close this issue. I might raise a PR with packer scripts to argument the cool stuff that you did because scaling out to tons of GPU does not work very nicely given NVIDIA download does not support massive scale download  but it says about 15 mins of provisioning time.
